index_test.go

```diff
@@ -1356,3 +1357,96 @@ func TestCompoundMultiIndex_FromObject(t *testing.T) {
 		t.Fatalf("\ngot:  %+v\nwant: %+v\n", got, want)
 	}
 }
+
+func TestCompoundMultiIndex_FromObject_IndexCountAndUniquenessProperty(t *testing.T) {
+	indexPermutations := [][]string{
+		{"Foo", "Qux", "QuxEmpty"},
+		{"Foo", "QuxEmpty", "Qux"},
+		{"QuxEmpty", "Qux", "Foo"},
+		{"QuxEmpty", "Foo", "Qux"},
+		{"Qux", "QuxEmpty", "Foo"},
+		{"Qux", "Foo", "QuxEmpty"},
+	}
+
+	fn := func(o TestObject) bool {
+		for _, perm := range indexPermutations {
+			indexer := indexerFromFieldNameList(perm)
+			ok, vals, err := indexer.FromObject(o)
+			if err != nil {
+				t.Logf("err: %v", err)
+				return false
+			}
+			if !ok {
+				t.Logf("should be ok")
+				return false
+			}
+			if !assertAllUnique(t, vals) {
+				return false
+			}
+		}
+		return true
+	}
+	seed := time.Now().UnixNano()
+	t.Logf("Using seed %v", seed)
+	cfg := quick.Config{Rand: rand.New(rand.NewSource(seed))}
+	if err := quick.Check(fn, &cfg); err != nil {
+		t.Fatalf("property not held: %v", err)
+	}
+}
+
+func assertAllUnique(t *testing.T, vals [][]byte) bool {
+	t.Helper()
+	s := make(map[string]struct{}, len(vals))
+	for _, index := range vals {
+		s[string(index)] = struct{}{}
+	}
+
+	if l := len(s); l != len(vals) {
+		t.Logf("expected %d unique indexes, got %v", len(vals), l)
+		return false
+	}
+	return true
+}
+
+func indexerFromFieldNameList(keys []string) *CompoundMultiIndex {
+	indexer := &CompoundMultiIndex{AllowMissing: true}
+	for _, key := range keys {
+		if key == "Foo" || key == "Baz" {
+			indexer.Indexes = append(indexer.Indexes, &StringFieldIndex{Field: key})
+			continue
+		}
+		indexer.Indexes = append(indexer.Indexes, &StringSliceFieldIndex{Field: key})
+	}
+	return indexer
+}
+
+func BenchmarkCompoundMultiIndex_FromObject(b *testing.B) {
+	obj := &TestObject{
+		ID:       "obj1-uuid",
+		Foo:      "Foo1",
+		Baz:      "yep",
+		Qux:      []string{"Test", "Test2"},
+		QuxEmpty: []string{"Qux", "Qux2"},
+	}
```

Comment:
I think this benchmark could be improved by using different random data for the values each time to see how different lengths of strings impacts the allocations, but this at least gives us some kind of baseline.